### PR TITLE
Bind secret nonces to signer keys and sync BIP327 nonce vectors

### PR DESCRIPTION
**Pull request by starius**
## Summary

This PR fixes [BIP327](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki) nonce handling in three related steps: binding secret nonces to the signer public key, matching optional [`NonceGen`](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#nonce-generation-1) input handling, and syncing the [BIP327 test vectors](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#test-vectors-and-reference-code) used by the test suite.

## Changes

- Bind `SecNonce` values to the signer public key and reject [`Sign`](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#signing) when a secret nonce is used with a different signer key.
- Preserve `UnknownKey` as the signing error when the signing key is not a member of the key aggregation context, even if the supplied `SecNonce` is also bound to a different key.
- Tighten `SecNonceBuilder::with_spices` so a spices seckey replaces a public key derived from an earlier seckey, while preserving an explicitly supplied public key.
- Match [BIP327 `NonceGen`](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#nonce-generation-1) optional inputs: when `sk` is absent, use raw random bytes as the nonce seed material; handle absent `aggpk`, `msg`, and `extra_in` as absent inputs; and include the `extra_in` length field as specified.
- Update the [BIP327 test vector](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#test-vectors-and-reference-code) fixtures and loaders for nonce generation, signature aggregation, and sign/verify so the test suite covers the current vector format and edge cases, including invalid partial signatures, invalid pubkeys, and invalid pubnonces from synced error vectors.

## Compatibility notes

This PR intentionally changes the public `SecNonce` representation to match [BIP327's 97-byte local secnonce form](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#nonce-generation-1), `k1 || k2 || pk`. The serialized size changes from 64 bytes to 97 bytes, `SEC_NONCE_SIZE` changes accordingly, and parsing now rejects the old 64-byte short form with a bad-length error.

`SecNonce::new` now requires the signer public key, and `SecNonce::random` now requires the signer public key. Callers that persisted or converted `SecNonce` as `[u8; 64]`, or constructed nonces without a public key, need to pass the public key that corresponds to the secret key that will later sign.

`SecNonceBuilder::build()` now requires either `with_pubkey` or `with_seckey` before `build()`. Calling `SecNonceBuilder::new(seed).build()` without a key now panics because [BIP327 `NonceGen`](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#nonce-generation-1) has mandatory `pk` input and the resulting local `SecNonce` must carry that key into [signing](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#signing).

`SecNoncePubkeyMismatch` is a local consistency check that catches accidental use of a secret nonce with the wrong signing key. It does not authenticate arbitrary raw `SecNonce` bytes supplied by another party.

## Security and compatibility

[BIP327 secret nonce generation](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#nonce-generation-1) includes the signer public key as protocol context and [BIP327 `Sign`](https://github.com/bitcoin/bips/blob/master/bip-0327.mediawiki#signing) checks that the public key embedded in `secnonce` matches the signing key. Carrying that binding inside `SecNonce` and checking it at signing time prevents a nonce generated for one signer key from being silently used with another signer key. The optional-input changes align Rust `NonceGen` behavior with BIP327 for callers that omit one or more optional inputs, including the case exercised by the updated nonce-generation vectors.

## Tests

This PR adds regression coverage for secret-nonce signer-key binding, signing-error precedence, `SecNonceBuilder::with_spices` pubkey merge behavior, and optional `NonceGen` inputs, then updates the BIP327 vector fixtures so those cases are exercised by the normal test suite. The vector harness now also validates synced signature-aggregation and sign/verify error cases rather than only carrying their updated metadata.

### CODE REVIEW

File: src/nonces.rs [hunk 1]
```diff
@@ -322,41 +345,45 @@ impl<'snb> SecNonceBuilder<'snb> {
...
+    ///
+    /// Panics if neither [`SecNonceBuilder::with_pubkey`] nor
+    /// [`SecNonceBuilder::with_seckey`] has been called.
     pub fn build(self) -> SecNonce {
-        let seckey_bytes = match self.seckey {
-            Some(seckey) => seckey.serialize(),
-            None => [0u8; 32],
+        let pubkey = self
+            .pubkey
+            .expect("BIP327 nonce generation requires a public key");
```
Comment (conduition) [inline src/nonces.rs]: I'd like to avoid panics as much as possible, especially because since this could easily cause existing downstream code using `musig2` to panic without compile-time warning. E.g. `SecNonce::build(seed).build()` would panic.

I can think of three options:

1. Return a `Result` from the `.build()` method.
2. Extend `SecNonceBuilder` with a generic parameter that restricts when the `.build()` method can be called.
3. Modify the `SecNonceBuilder::new` method to accept a public or secret key alongside the seed input.  Remove the `.with_pubkey(...)` method.

Option 2 is the only one which preserves some semblance of compatibility... but I think it's more effort than it's worth. It would complicate the code, and generate unclear compile errors downstream. 

I think I'd be OK with sacrificing API compatibility with this patch, so I lean towards option 3. I'm open to suggestions if you can think of anything better

File: src/nonces.rs [hunk 2]
```diff
@@ -307,10 +321,19 @@ impl<'snb> SecNonceBuilder<'snb> {
...
-    /// in `spices` if they are present.
+    /// The secret key and message will be merged with those in `spices`, preferring
+    /// parameters in `spices` if they are present. If a secret key is present in
+    /// `spices`, it will also derive the signing public key unless this builder
+    /// already has an explicitly provided public key.
     pub fn with_spices(mut self, spices: SecNonceSpices<'snb>) -> SecNonceBuilder<'snb> {
-        self.seckey = spices.seckey.or(self.seckey);
+        if let Some(seckey) = spices.seckey {
+            self.seckey = Some(seckey);
+            if self.pubkey.is_none() || !self.pubkey_is_explicit {
```
Comment (conduition) [inline src/nonces.rs]: Interesting, I like this. Basically if the caller has explicitly asked to use a certain public key as a salt, we should honor that, even if given a mismatching secret key later.

We should also extend the `.with_seckey(...)` method to follow this behavior now that pubkeys are a mandatory input to the nonce builder.

File: src/nonces.rs [hunk 3]
```diff
@@ -322,41 +345,45 @@ impl<'snb> SecNonceBuilder<'snb> {
...
-        // fashion as the aggregated pubkey.
-        match self.pubkey {
-            None => hasher.update([0]),
-            Some(pubkey) => {
-                hasher.update([33]); // individual pubkey len
-                hasher.update(pubkey.serialize());
-            }
-        }
+        hasher.update([33]); // individual pubkey len
+        hasher.update(pubkey.serialize());
```
Comment (conduition) [inline src/nonces.rs]:
```suggestion
            .chain_update(rand)
            .chain_update([33]) // individual pubkey len
            .chain_update(pubkey.serialize());
```

Comment (conduition) [reply src/nonces.rs]: Wow, github butchered the rendering of that suggestion. Look at the line numbers for a more accurate delta

File: src/test_vectors/sign_verify_vectors.json [hunk 4]
```diff
@@ -196,7 +196,7 @@
             "comment": "Invalid pubnonce"
         },
         {
-            "sig": "68537CC5234E505BD14061F8DA9E90C220A181855FD8BDB7F127BB12403B4D3B",
```
Comment (conduition) [inline src/test_vectors/sign_verify_vectors.json:199]: This explains why my vectors weren't matching, they changed after this crate was published and I didn't notice to update them: https://github.com/bitcoin/bips/commit/508e3a6a40a6e73c73cbfa8a33aa18a2bc7b9d91

File: src/nonces.rs (hunk 1 above)
Comment (conduition) [reply src/nonces.rs]: Fixed by https://github.com/starius/musig2/pull/1

File: src/nonces.rs (hunk 2 above)
Comment (conduition) [reply src/nonces.rs]: Fixed by https://github.com/starius/musig2/pull/1

File: src/nonces.rs (hunk 3 above)
Comment (conduition) [reply src/nonces.rs]: Fixed in https://github.com/starius/musig2/pull/1

**Comment by conduition** [comment]
@starius I have submitted some suggestions in https://github.com/starius/musig2/pull/1, which targets your `nonce-fixes` branch (the source of this PR). Please merge if you like the changes :)

**Comment by starius** [comment]
I fixed the record in CHANGELOG: removed duplicate "now".

**Comment by conduition** [comment]
Taking second look at the overall changes now...

**Review by conduition — APPROVED** [review]
LGTM 👍

**Comment by conduition** [comment]
Thanks for your diligent work @starius :heavy_check_mark: :100:

**Comment by storopoli** [comment]
@conduition any chance that we can backport this to the `0.1.x` branch and have a new release on that branch?

Most of downstream users (at least two cases in Alpen Labs) will use `musig2` alongside `bitcoin`. Latest `bitcoin` (0.32.9) is on `secp256k1` 0.29.1 whereas `musig2` is on `0.31.x` this makes impossible to reconcile these two crates.

**Comment by conduition** [comment]
Hi @storopoli, dang that's a shame the rust-bitcoin folks still haven't upgraded. Looks like there is a renewed push for it now after the latest version of `secp256k1` exposes musig APIs: https://github.com/rust-bitcoin/rust-bitcoin/issues/6117

Since `musig2` is mainly targeted at Bitcoin usage, i'm also weighing the option to just peg `secp`'s dependency on `secp256k1` to whatever the `bitcoin` crate is using. I hope they will eventually upgrade but i'm getting a bit tired of waiting.

**Comment by conduition** [comment]
@storopoli OK, i have just completed the backporting: https://github.com/conduition/musig2/commit/d2870d2f2ba0cc05f85af46b028fded17130050d

These changes have been published as `musig2@0.1.2`

**Comment by storopoli** [comment]
> @storopoli OK, i have just completed the backporting: [d2870d2](https://github.com/conduition/musig2/commit/d2870d2f2ba0cc05f85af46b028fded17130050d)
> 
> These changes have been published as `musig2@0.1.2`

Thank you!
